aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/MilestoneStepTest.java

```diff
@@ -0,0 +1,166 @@
+
+    @Rule
+    public RestartableJenkinsRule story = new RestartableJenkinsRule();
+
+    @Before
+    public void clear() {
+        MilestoneStepExecution.clear();
+    }
+
+    @Test
+    public void buildsMustPassThroughInOrder() {
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                p.setDefinition(new CpsFlowDefinition(
+                        "semaphore 'inorder'\n" +
+                        "echo 'Before milestone'\n" +
+                        "milestone()\n" +
+                        "echo 'Passed first milestone'\n" +
+                        "milestone()\n" +
+                        "echo 'Passed second milestone'\n" +
+                        "milestone()\n" +
+                        "echo 'Passed third milestone'\n"));
+                WorkflowRun b1 = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("inorder/1", b1);
+                WorkflowRun b2 = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("inorder/2", b2);
+
+                // Let #2 continue so it finish before #1
+                SemaphoreStep.success("inorder/2", null);
+                story.j.assertBuildStatusSuccess(story.j.waitForCompletion(b2));
+
+                // Let #1 continue, so it must be early cancelled since #2 already went into milestone 1
+                SemaphoreStep.success("inorder/1", null);
+                story.j.assertBuildStatus(Result.NOT_BUILT, story.j.waitForCompletion(b1));
+                story.j.assertLogNotContains("Passed first milestone", b1);
+            }
+        });
+    }
+
+    @Test
+    public void olderBuildsMustBeCancelledOnMilestoneExit() {
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                p.setDefinition(new CpsFlowDefinition(
+                        "milestone()\n" +
+                        "echo 'First milestone'\n" +
+                        "semaphore 'wait'\n" +
+                        "milestone()\n" +
+                        "echo 'Second milestone'\n"));
+                WorkflowRun b1 = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("wait/1", b1);
+                WorkflowRun b2 = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("wait/2", b2);
+                // Now both #1 and #2 are into milestone 1
+
+                // Let #2 continue so it leaves milestone 1
+                SemaphoreStep.success("wait/2", null);
+                story.j.waitForCompletion(b2);
+
+                // Once #2 continues and reaches milestone 2 then #1 is automatically cancelled
+                // TODO: determine why it's getting ABORTED instead of NOT_BUILT
```

Comment:
Yes. The cause of this is that a call to `Run#getExecutor().interrupt(Result, CauseOfInterruption ... causes)` in a `WorkflowRun` in turn calls `AsynchronousExecution#interrupt(boolean)` which calls the specific implementation of `AsynchronousExecution` in `WorkflowRun`, which  finally calls `execution.interrupt(Result.ABORTED);`, so the initial `Result` is omitted.

The above behavior is a bug IMO, but not fixable without API changes I think.

Anyway it's locally avoidable by calling `((FlowExecutionOwner.Executable) r).asFlowExecutionOwner().get().interrupt(Result, CauseOfInterruption ...);` directly in `MilestoneStepExecution`. Fixing...